### PR TITLE
Duplicate Ways Check

### PR DESCRIPTION
**Pull request by savannahostrowski**
The Duplicate Ways check flags Ways in OpenStreetMap which are either partially or completely duplicated.

This pull request covers the logic, unit tests, documentation, and configuration for the Duplicate Ways check.

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    public static final String DUPLICATE_EDGE_INSTRUCTIONS = "This way, {0,number,#}, "
+            + "has at least one duplicate segment. ";
+
+    public static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    public static final int ZERO_LENGTH = 0;
+    public static final String AREA_KEY = AreaTag.KEY;
+
+    // A map of segments and list of Edge identifiers
+    private final Map<Segment, Set<Long>> globalSegments = new HashMap<>();
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Is some cases, this data structure has a potential to fill the memory, cause OOM exceptions. I'd suggest finding a less memory-intensive approach to find duplicate ways.

**Comment by savannahostrowski** [comment]
I've refactored the logic for this check based on @mertemin's comments surrounding memory usage. Using a HashSet within the flag method scope(as opposed to the original global HashMap), I think this is more along the lines of what we're looking for. I've also added in another preliminary filtering conditional to ensure that we are only checking master Edges in the Duplicate Ways check.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+{
+    // You can use serialver to regenerate the serial UID.
+    private static final long serialVersionUID = 1L;
+    public static final String DUPLICATE_EDGE_INSTRUCTIONS = "This way, {0,number,#}, "
+            + "has at least one duplicate segment.";
+
+    public static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    public static final int ZERO_LENGTH = 0;
+    public static final String AREA_KEY = AreaTag.KEY;
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: These statics don't have to be exposed publicly from the class.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge
+                // Check to see that the Edge is a master Edge
+                && ((Edge) object).isMasterEdge()
+                // Check to see that the edge has not already been seen
+                && !this.isFlagged(((Edge) object).getMasterEdgeIdentifier())
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: If this condition is being checked, then `((Edge) object).isMasterEdge()` must be true, as well as `object instanceof Edge`. So you don't have to cast `object` here. Just get it's identifer and check if it is flagged or not.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        super(configuration);
+    }
+
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return object instanceof Edge
+                // Check to see that the Edge is a master Edge
+                && ((Edge) object).isMasterEdge()
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Please consider using [`Edge.isMasterEdgeIdentifier` method](https://github.com/osmlab/atlas/blob/1dbcbe5818d018f7c919b51d8ea82902ec1a5fa7/src/main/java/org/openstreetmap/atlas/geography/atlas/items/Edge.java#L30).

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java (hunk 2 above)
Comment (mertemin) [reply src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Using `AreaTag.KEY` might be cleaner than using `AREA_KEY`.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+public class DuplicateWaysCheck extends BaseCheck
+{
+    // You can use serialver to regenerate the serial UID.
+    private static final long serialVersionUID = 1L;
+    public static final String DUPLICATE_EDGE_INSTRUCTIONS = "This way, {0,number,#}, "
+            + "has at least one duplicate segment.";
+
+    public static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    public static final int ZERO_LENGTH = 0;
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Please reuse [`Distance.ZERO`](https://github.com/osmlab/atlas/blob/1dbcbe5818d018f7c919b51d8ea82902ec1a5fa7/src/main/java/org/openstreetmap/atlas/utilities/scalars/Distance.java#L24).

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        // Out the non-master Edges as the bounds Edges are not guaranteed to be uni-directional
+        final Iterable<Edge> edgesInBounds = Iterables
+                .stream(edge.getAtlas().edgesIntersecting(bounds)).filter(Edge::isMasterEdge);
+
+        for (final Edge edgeInBounds : edgesInBounds)
+        {
+            // If the Edge found in the bounds has an area tag or if the Edge has a length of 0
+            // Or if the Edge has already been flagged before then continue because we don't want to
+            // Flag area Edges or duplicate Nodes
+            if (edgeInBounds.getTags().containsKey(AREA_KEY)
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: `getTags` is fetching all the tags. Please rely on `getTag` method.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                    || edgeInBounds.asPolyLine().length().equals(Distance.meters(ZERO_LENGTH))
+                    || this.isFlagged(edgeInBounds.getMasterEdgeIdentifier()))
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: `edgeInBounds` is verified that it is master edge on line 73. No need to call `getMasterEdgeIdentifier`.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            {
+                continue;
+            }
+
+            // If the Set of Edges does not contain the Edge found in the bounds
+            if (!allEdgeSegments.containsAll(edgeInBounds.asPolyLine().segments()))
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Please consider using [`PolyLine.intersects` method](https://github.com/osmlab/atlas/blob/c0965af0fd1563844fff505a2a80cc79b95f655e/src/main/java/org/openstreetmap/atlas/geography/PolyLine.java#L592) to check for an overlap. You can also utilize [`PolyLine.intersections` method](https://github.com/osmlab/atlas/blob/c0965af0fd1563844fff505a2a80cc79b95f655e/src/main/java/org/openstreetmap/atlas/geography/PolyLine.java#L570) to get overlapping locations.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+                }
+                // Add all Segments flattened to the Set of Segments
+                allEdgeSegments.addAll(edgeInBounds.asPolyLine().segments());
+            }
+            // A full duplicate Edge was found in the Set of allEdgePolyLines
+            else
+            {
+                this.markAsFlagged(edgeInBounds.getMasterEdgeIdentifier());
+                return Optional.of(this.createFlag(edgeInBounds,
+                        this.getLocalizedInstruction(0, edgeInBounds.getOsmIdentifier())));
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: A full overlap has the same instruction. I'd suggest not worrying about the full overlap. Just flag whenever an overlap is found. Atlas has edges vs OSM has ways. Edges by definition are partial OSM ways, so full overlap or not, you'll almost always flag partial OSM ways anyways.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Set<Segment> allEdgeSegments = new HashSet<>();
+        final Edge edge = (Edge) object;
+
+        final Rectangle bounds = edge.asPolyLine().bounds();
+        // Get Edges which are contained by or intersect the bounds, and then filter
+        // Out the non-master Edges as the bounds Edges are not guaranteed to be uni-directional
+        final Iterable<Edge> edgesInBounds = Iterables
+                .stream(edge.getAtlas().edgesIntersecting(bounds)).filter(Edge::isMasterEdge);
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: You can possibly say `final Iterable<Edge> edgesInBounds = edge.getAtlas().edgesIntersecting(bounds, Edge::isMasterEdge)`. `edgesIntersecting` optionally accepts a `Predicate`.

**Comment by savannahostrowski** [comment]
Additionally, for my own knowledge, was the concern here mostly just to do with the use of the HashMap to store Segments and associated Edge IDs as a global variable?  Or is the concern more specific to how Atlas Checks are being run (is it being run on a country-by-country/region-by-region basis or is the whole world's data being processed in one fell swoop? 😄 )

**Comment by mertemin** [comment]
@savannahostrowski Atlas-checks is by design flexible to accept any given part of the world to execute (or the whole world for that matter). The concern with accumulating state and using it as you go validate more features is that results will be determined by the order of features processed and memory footprint of a such solution could be big (based on the area of interest). Atlas could be utilized and possibly overlapping edges could be retrieved without additional memory tracking.

Here is an example:
edge-x (A B C D)
edge-y (H I K L)
edge-z (A B C)
edge-t (C D)

If you process x, y, z and finally t, then you'd flag z and t for overlap. If you process, t, z, y and finally x, then x will be flagged only.

**Comment by savannahostrowski** [comment]
Addressed @mertemin comments regarding reuse of constants, and fine tuning preliminary filtering. Opted to use [.overlapsShapeOf()](https://github.com/osmlab/atlas/blob/c0965af0fd1563844fff505a2a80cc79b95f655e/src/main/java/org/openstreetmap/atlas/geography/PolyLine.java#L830) to detect duplicate Ways. Let me know if this still isn't sufficient!

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+public class DuplicateWaysCheck extends BaseCheck
+{
+    // You can use serialver to regenerate the serial UID.
+    private static final long serialVersionUID = 1L;
+    private static final String DUPLICATE_EDGE_INSTRUCTIONS = "This way, {0,number,#}, "
+            + "has at least one duplicate segment.";
+
+    private static final List<String> FALLBACK_INSTRUCTIONS = Arrays
+            .asList(DUPLICATE_EDGE_INSTRUCTIONS);
+    private static final Distance ZERO_LENGTH = Distance.ZERO;
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Consider using `ZERO_DISTANCE` to be consistent with the type.

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        return object instanceof Edge
+                // Check to see that the Edge is a master Edge
+                && Edge.isMasterEdgeIdentifier(object.getIdentifier())
+                // Check to see that the edge has not already been seen
+                && !this.isFlagged(object.getIdentifier())
+                // Check to see that the edge is car navigable
+                && HighwayTag.isCarNavigableHighway(object)
+                // The edge is not part of an area
+                && !object.getTags().containsKey(AreaTag.KEY);
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: Consider using `object.getTag(AreaTag.KEY)` to be consistent (see line #77).

File: src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                continue;
+            }
+
+            final PolyLine edgeInBoundsPoly = edgeInBounds.asPolyLine();
+
+            // Checks that either the edgeInBounds PolyLine overlaps the edgePoly or
+            // That the edgePoly overlaps the edgeInBoundsPoly
+            if (edgeInBoundsPoly.overlapsShapeOf(edgePoly)
+                    || edgePoly.overlapsShapeOf(edgeInBoundsPoly))
```
Comment (mertemin) [inline src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java]: `overlapsShapeOf` is comparing individual segments. If there is overlap, then that would be found with the first call. If there is no overlap, then it wouldn't be found in both overlap checks. I'd suggest dropping one of the overlap calls, as it seems redundant.
